### PR TITLE
throw IOException instead of InvalidProtocolBufferException when appropriate

### PR DESCRIPTION
**Pull request by jhump**
When the `Parser` type and generated parsing constructors were introduced in v2.5, some functionality became broken.

Prior to that change, if a message could not be read due to an `IOException` on an underlying I/O stream, then the thrown exception would be that same `IOException`. The sub-class `InvalidProtocolBufferException` was thrown when a coded input stream detected invalid message encoding.

Being able to distinguish between these two cases is very useful in servers that read protobuf requests. Socket exceptions, especially read timeouts, are very different from incorrect encoding. For example, in the case of a read timeout (where the socket hasn't closed, and a client still wants a response), being able to reply with a 408 instead of a 400 is invaluable (since clients will typically treat 400 as a permanent failure and won't retry). Even in server-side metrics, being able to record distinct counts of socket exceptions vs. bad requests is useful.

This pull request restore the ability to distinguish. The fix is a little bit ugly because everything now goes through `Parser`, and its methods only declare `InvalidProtocolBufferException` as the thrown type. So the `Parser` methods must wrap `IOException`s in `InvalidProtocolBufferException`s (since changing throws clause would break backwards compatibility). Other methods, that are still declared to throw `IOException` (like generated `Message.parseFrom(InputStream)` and `Builder.mergeFrom(InputStream)`) will handle unwrapping, thus restoring their behavior to pre-2.5 form.

[Review on Reviewable](https://reviewable.io/reviews/google/protobuf/896)

### CODE REVIEW

**Comment by jhump** [comment]
cc @cconroy

**Comment by xfxyjwf** [comment]
@pherl

**Comment by pherl** [comment]
Reviewed 3 of 4 files at r1.
Review status: 3 of 4 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

<sup>**[src/google/protobuf/compiler/java/java_message.cc, line 669 \[r1\]](https://reviewable.io:443/reviews/google/protobuf/896#-K1GZo-0pHi0FErUwel5)** ([raw file](https://github.com/google/protobuf/blob/ec9d488418249c3afdf32c44002f5a7528f58047/src/google/protobuf/compiler/java/java_message.cc#L669)):</sup>
I would leave those methods unchanged. Clients should also be able to check the getCause() by themselves to check. Also the behavior is more consistent with PARSER, and it won't increase generated code size.


---

<sup>**[src/google/protobuf/compiler/java/java_message_builder.cc, line 541 \[r1\]](https://reviewable.io:443/reviews/google/protobuf/896#-K1G_JrXYTPgA0jZpaEV)** ([raw file](https://github.com/google/protobuf/blob/ec9d488418249c3afdf32c44002f5a7528f58047/src/google/protobuf/compiler/java/java_message_builder.cc#L541)):</sup>
same here.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/google/protobuf/896)

**Comment by jhump** [comment]
Review status: 3 of 4 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

<sup>**[src/google/protobuf/compiler/java/java_message.cc, line 669 \[r1\]](https://reviewable.io:443/reviews/google/protobuf/896#-K1GZo-0pHi0FErUwel5)** ([raw file](https://github.com/google/protobuf/blob/ec9d488418249c3afdf32c44002f5a7528f58047/src/google/protobuf/compiler/java/java_message.cc#L669)):</sup>
These methods are already spec'ed to throw `IOException`, which means callers already have to catch/expect `IOException` instead of `InvalidProtocolBufferException`. So this behavior still adheres to existing spec.

Furthermore, this behavior is *far* more intuitive and easier to use. Making callers unwrap an exception is just bad API. Instead, this change allows idiomatic Java: multiple catch blocks that switch on type.

What do you think if I also clarify the spec, by adding `@throws IOException` and `@throws InvalidProtocolBufferException` java-doc for the impacted methods. That maintains compatibility with the current spec while improving it and clarifying it for future usages.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/google/protobuf/896)

**Comment by jhump** [comment]
ping @xfxyjwf @pherl

**Comment by pherl** [comment]
Review status: 3 of 4 files reviewed at latest revision, 2 unresolved discussions.

---

<sup>**[src/google/protobuf/compiler/java/java_message.cc, line 669 \[r1\]](https://reviewable.io:443/reviews/google/protobuf/896#-K1GZo-0pHi0FErUwel5)** ([raw file](https://github.com/google/protobuf/blob/ec9d488418249c3afdf32c44002f5a7528f58047/src/google/protobuf/compiler/java/java_message.cc#L669)):</sup>
The only concern I have is the additional code generated for each class. There are lots of concerns of generated code and PermGen bloat. It would be fine if you could abstract this into some internal util functions in the runtime and delegate to the runtime function to minimize the codegen impact.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/google/protobuf/896)

**Comment by jhump** [comment]
<sup>**[src/google/protobuf/compiler/java/java_message.cc, line 669 \[r1\]](https://reviewable.io:443/reviews/google/protobuf/896#-K1GZo-0pHi0FErUwel5)** ([raw file](https://github.com/google/protobuf/blob/ec9d488418249c3afdf32c44002f5a7528f58047/src/google/protobuf/compiler/java/java_message.cc#L669)):</sup>
Ok. I've changed the code so the generated code should be equivalent number of byte code instructions in message classes and just one extra method invocation in the builder classes (to call `InvalidProtocolBufferException#unwrapIOException`).

I also realized that the test I forgot to `git add` the test I wrote. So now there's a test!


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/google/protobuf/896)

**Comment by jhump** [comment]
<sup>**[src/google/protobuf/compiler/java/java_message.cc, line 669 \[r1\]](https://reviewable.io:443/reviews/google/protobuf/896#-K1GZo-0pHi0FErUwel5)** ([raw file](https://github.com/google/protobuf/blob/ec9d488418249c3afdf32c44002f5a7528f58047/src/google/protobuf/compiler/java/java_message.cc#L669)):</sup>
I just realized that the byte code count should be the same. The one extra method invocation in generated builder classes is offset by one less method invocation in message classes since we no longer invoke `IOException#getMessage` when throwing the `InvalidProtocolBufferException`  :)


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/google/protobuf/896)

**Comment by jhump** [comment]
Another ping @pherl. I think I addressed your comments. Can you take another look?

**Comment by pherl** [comment]
Sorry for the delay. Loos good in general. Only a minor thing, please see inline.


---

Reviewed 6 of 6 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

<sup>**[src/google/protobuf/compiler/java/java_message.cc, line 667 \[r2\]](https://reviewable.io:443/reviews/google/protobuf/896#-K2x9_UooruLXARCtssD)** ([raw file](https://github.com/google/protobuf/blob/9ea0d5b767caebe0f526e3a833cb483ae325a459/src/google/protobuf/compiler/java/java_message.cc#L667)):</sup>
Use com.google.protobuf.GeneratedMessage.parseWithIOException for a more explicit static method call. Ditto to the rest of the file.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/google/protobuf/896)

**Comment by jhump** [comment]
Sorry it took me so long to revisit this. I made the requested change.

---







*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/google/protobuf/896#-:-K7Mxis1hvDH32g9Not9:1721418522)*

**Comment by jhump** [comment]
CI won't go green because of unrelated issue: same test that is red in this PR is broken on master according to Travis CI.

Is someone looking into getting master green? Should I rebase back to a point before that issue was introduced or just wait for master to be fixed and then rebase to that?

**Comment by xfxyjwf** [comment]
The python_cpp test is known to be broken. I think it's fine since this pull request is about Java. Can you rebase this PR on the current master? We just did some refactoring of the Java code.

**Comment by jhump** [comment]
@xfxyjwf: After rebasing after the refactoring in Java code, there are now more red tests -- *all* of which are also broken on master. One of them is Java, but it is red on master with the same error message (and furthermore looks like it can't be related to my changes).

Any advice on how I should proceed? Or can you merge this branch despite the errors?

File: java/src/test/java/com/google/protobuf/ParseExceptionsTest.java [hunk 1]
```diff
@@ -0,0 +1,211 @@
+package com.google.protobuf;
```
Comment (xfxyjwf) [inline java/src/test/java/com/google/protobuf/ParseExceptionsTest.java]: Move this to java/core/test/... ?

**Comment by xfxyjwf** [comment]
@jhump I just merged https://github.com/google/protobuf/pull/1143 to fix these Java/C++ broken tests. Sorry for that, but could you update your commits based on the latest master? I also noticed a test file still in the old location.

File: java/src/test/java/com/google/protobuf/ParseExceptionsTest.java (hunk 1 above)
Comment (jhump) [inline java/src/test/java/com/google/protobuf/ParseExceptionsTest.java]: Doh! Good catch.

There were no conflicts after rebase and `mvn test` ran clean, so I totally didn't notice how this didn't fit with the rearranged modules. Sorry!

(Fixed now)

**Comment by jhump** [comment]
@xfxyjwf This is rebased, all comments addressed, and the only failing test is the known-failing `python_cpp`. I think it's finally ready to land.

**Comment by xfxyjwf** [comment]
Thanks!

**Pull request merged**